picasso/src/androidTest/java/com/squareup/picasso3/PicassoComposePainterTest.kt

```diff
@@ -0,0 +1,148 @@
+import androidx.compose.foundation.layout.requiredSize
+import androidx.compose.runtime.CompositionLocalProvider
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.onSizeChanged
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.test.junit4.createComposeRule
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.dp
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.platform.app.InstrumentationRegistry
+import com.google.common.truth.Truth.assertThat
+import com.squareup.picasso3.compose.rememberPainter
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class PicassoComposePainterTest {
+
+    @get:Rule
+    val rule = createComposeRule()
+
+    @Test
+    fun firstFrameConsumesStateFromLayout() {
+        lateinit var lastRequest: Request
+        val context = InstrumentationRegistry.getInstrumentation().targetContext
+        val picasso = Picasso.Builder(context)
+            .addRequestHandler(object : RequestHandler() {
+                override fun canHandleRequest(data: Request): Boolean = true
+                override fun load(picasso: Picasso, request: Request, callback: Callback) {
+                    lastRequest = request
+                }
+            })
+            .build()
+        var size: IntSize by mutableStateOf(IntSize.Zero)
+        var initialFrame = true
+
+        rule.setContent {
+            CompositionLocalProvider(LocalDensity provides Density(1f)) {
+                val painter = picasso.rememberPainter {
+                    it.load("http://example.com/")
+                        .addHeader("width", size.width.toString())
+                        .addHeader("height", size.height.toString())
+                }
+                Canvas(
+                    Modifier
+                        .requiredSize(9.dp)
+                        .onSizeChanged { size = it }
+                ) {
+                    val canvasSize = this.size
+
+                    if (initialFrame) {
+                        initialFrame = false
+                        // Initial request was made with uninitialized size.
+                        assertThat(lastRequest.headers?.toMultimap()).containsAtLeastEntriesIn(
+                            mapOf(
+                                "width" to listOf("0"),
+                                "height" to listOf("0"),
+                            )
+                        )
+                    }
+
+                    with(painter) {
+                        draw(canvasSize)
+                    }
+
+                    // Draw triggers request was made with size.
+                    assertThat(lastRequest.headers?.toMultimap()).containsAtLeastEntriesIn(
```

Comment:
I don't think this will actually pass right now because it's not going to be ran synchronously? Might need to warm the memory cache with the fake request?